File: leetcode2958/leetcode2958.py
from collections import Counter, defaultdict

class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        ans = 0
        left = 0
        cnt = defaultdict(int)
        for right in range(len(nums)):
            # _sum += nums[right]
            cnt[nums[right]] += 1
            while cnt[nums[right]] > k:
                # _sum -= nums[left]
                cnt[nums[left]] -= 1
                if cnt[nums[left]] == 0:
                    cnt.pop(nums[left])
                left += 1
            ans = max(ans, right - left + 1)
        return ans

File: leetcode2958/leetcode2958.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    public:
        int maxSubarrayLength(vector<int>& nums, int k) {
            int ans = 0, left = 0;
            unordered_map<int, int> cnt;
            for (int right = 0; right < nums.size(); right ++){
                cnt[nums[right]] ++;
                while (cnt[nums[right]] > k)
                {
                    cnt[nums[left]] --;
                    if (cnt[nums[left]] == 0) cnt.erase(nums[left]);
                    left ++;
                }
                ans = max(ans, right - left + 1);
            }
            return ans;
        }
    };